rpc/block_reload.c

```diff
@@ -0,0 +1,421 @@
+    }
+    goto out;
+  }
+
+  if (GB_ALLOC(info) < 0) {
+    errCode = ENOMEM;
+    goto out;
+  }
+
+  ret = blockGetMetaInfo(glfs, blk->block_name, info, NULL);
+  if (ret) {
+    errCode = ret;
+    goto out;
+  }
+
+/*
+  if (!blk->force) {
+    list = glusterBlockGetListFromInfo(info);
+    if (!list) {
+      errCode = ENOMEM;
+      goto out;
+    }
+
+    errCode = glusterBlockCheckCapabilities((void *)blk, DELETE_SRV, list, NULL, &errMsg);
+    if (errCode) {
+      LOG("mgmt", GB_LOG_ERROR,
+          "glusterBlockCheckCapabilities() for block %s on volume %s failed",
+          blk->block_name, blk->volume);
+      goto out;
+    }
+  }
+*/
+
+  GB_STRCPYSTATIC(robj.block_name, blk->block_name);
+  GB_STRCPYSTATIC(robj.gbid, info->gbid);
+
+  errCode = glusterBlockReloadRemoteAsync(info, glfs, &robj, &savereply);
+  if (errCode) {
+    LOG("mgmt", GB_LOG_WARNING, "glusterBlockReloadRemoteAsync: return %d "
+        "on block %s for volume %s", errCode, blk->block_name, blk->volume);
+  }
+
+ out:
+  GB_METAUNLOCK(lkfd, blk->volume, errCode, errMsg);
+  blockFreeMetaInfo(info);
+//  blockServerDefFree(list);
+
+ optfail:
+  LOG("mgmt", ((!!errCode) ? GB_LOG_ERROR : GB_LOG_INFO),
+      "reload cli return %s, volume=%s blockname=%s",
+      errCode ? "failure" : "success", blk->volume, blk->block_name);
+
+  if (lkfd && glfs_close(lkfd) != 0) {
+    LOG("mgmt", GB_LOG_ERROR,
+        "glfs_close(%s): for block %s on volume %s failed[%s]",
+        GB_TXLOCKFILE, blk->block_name, blk->volume, strerror(errno));
+  }
+
+  blockReloadCliFormatResponse(blk, errCode, errMsg, savereply, reply);
+  LOG("cmdlog", ((!!errCode) ? GB_LOG_ERROR : GB_LOG_INFO), "%s",
+      reply ? reply->out : "*Nil*");
+
+  if (savereply) {
+    GB_FREE(savereply->d_attempt);
+    GB_FREE(savereply->d_success);
+    GB_FREE(savereply);
+  }
+  GB_FREE(errMsg);
+
+  return reply;
+}
+
+
+static blockResponse *
+block_reload_1_svc_st(blockReload *blk, struct svc_req *rqstp)
+{
+  char *exec = NULL;
+  blockResponse *reply = NULL;
+
+
+  LOG("mgmt", GB_LOG_INFO,
+      "reload request, blockname=%s filename=%s", blk->block_name, blk->gbid);
+
+  if (GB_ALLOC(reply) < 0) {
+    goto out;
+  }
+  reply->exit = -1;
+
+  if (GB_ASPRINTF(&exec, "targetcli %s target=%s%s storage_object=%s", GB_RELOAD,
+                  GB_TGCLI_IQN_PREFIX, blk->gbid, blk->block_name) == -1) {
+    goto out;
+  }
+
+  if (GB_ALLOC_N(reply->out, 8192) < 0) {
+    GB_FREE(reply);
+    goto out;
+  }
+
+  /* currently targetcli throw few error msgs for keywords in saveconfig.json
+   * which are unknown and also return few warning msgs and error code, hence
+   * not using GB_CMD_EXEC_AND_VALIDATE() macro and not caring for return
+   * value for now. For now lets just run the restoreconfig command and use
+   * blockCheckBlockLoadedStatus() to check we get success.
+   */
+  gbRunner(exec);
+
+  if (blockCheckBlockLoadedStatus(blk->block_name, blk->gbid, reply)) {
```

Comment:
Maybe we should check the load status before running the gbRunner() ? And for the ones which are already loaded, we can just skip it with success.
Will this make sense ?

Reply:
@lxbsz this is reload feature, think it like systemctl restart, it should definitely reload, what do you think ?

Reply:
@pkalever 
BTW, in case if the target/iqn are both already loaded successfully, and here we run the `targetcli reaload xxx`, will there be any strange thing to happen ? If the targetcli could work well without any complaining I am okay with this.